scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/data/TaskProActiveDataspaces.java

```diff
@@ -285,162 +318,241 @@ protected void logDataspacesStatus(String message, DataspacesStatusLevel level)
-                                        DataspacesStatusLevel.ERROR);
-                            }
-                            return true;
-                        }
-                    }));
+    /*
+     * Create the folder hierarchy and select the files to copy
+     * from the specified list of FileObjects.
+     */
+    private void createFolderHierarchySequentially(
+            DataSpacesFileObject destination, String spaceUri, List<DataSpacesFileObject> spaceFiles,
+            Map<String, DataSpacesFileObject> filesToCopy) throws FileSystemException {
 
-                }
-                relPathes.add(relativePath);
+        for (DataSpacesFileObject fileObject : spaceFiles) {
+            String relativePath = relativize(spaceUri, fileObject);
+
+            DataSpacesFileObject target = destination.resolveFile(relativePath);
+
+            if (target.isFolder()) {
+                target.createFolder();
+            } else if (target.isFile()) {
+                target.getParent().createFolder();
             }
 
-            StringBuilder exceptionMsg = new StringBuilder();
-            String nl = System.lineSeparator();
-            for (Future<Boolean> f : transferFutures) {
-                try {
-                    f.get();
-                } catch (InterruptedException | ExecutionException e) {
-                    logger.error("", e);
-                    exceptionMsg.append(StackTraceUtil.getStackTrace(e)).append(nl);
-                }
+            DataSpacesFileObject oldFileObject = filesToCopy.put(relativePath, fileObject);
+            if (oldFileObject != null) {
+                String message = fileObject.getRealURI()
+                        + " will be copied instead of " + oldFileObject.getRealURI() + ".\n "
+                        + "Precedence order is output space, input space, user space, global space.";
+                logger.warn(message);
+                logDataspacesStatus(message, DataspacesStatusLevel.WARNING);
             }
-            if (exceptionMsg.length() > 0) {
-                toBeThrown = new FileSystemException(
-                        "Exception(s) occurred when transferring input files : " + nl + exceptionMsg.toString());
+        }
+    }
+
+    private void handleResults(List<Future<Boolean>> transferFutures) throws FileSystemException {
+
+        StringBuilder message = new StringBuilder();
+        String nl = System.lineSeparator();
+
+        for (Future<Boolean> future : transferFutures) {
+            try {
+                future.get();
+            } catch (InterruptedException | ExecutionException e) {
+                logger.error("Exception while fetching dataspace transfer result ", e);
+                message.append(StackTraceUtil.getStackTrace(e)).append(nl);
             }
+        }
 
-            if (toBeThrown != null) {
-                throw toBeThrown;
+        if (message.length() > 0) {
```

Comment:
Thank you Marco but I have already received 2 reviews and the changes are merged since 4 days. To answer to your question, it was handled like that before my changes. I have kept this piece of code because it's really easy to introduce a regression with the current flow of execution.